ui_components/src/village/VillageUpgrades.js

```diff
@@ -0,0 +1,754 @@
+import { apiFetch } from "../utils/network.js";
+import { useModal } from '../utils/modalContext.js';
+
+export function VillageUpgrades({
+    playerSeatState,
+    villageName,
+    villageAPI,
+    buildingUpgradeDataState,
+    setBuildingUpgradeDataState,
+    resourceDataState,
+    setProposalDataState,
+}) {
+    const [selectedBuilding, setSelectedBuilding] = React.useState(null);
+    const [selectedUpgrade, setSelectedUpgrade] = React.useState(null);
+    const [hoveredUpgrade, setHoveredUpgrade] = React.useState(null);
+    const { openModal } = useModal();
+    const current_materials = resourceDataState.find(resource => resource.resource_name == "materials").count;
+    const current_food = resourceDataState.find(resource => resource.resource_name == "food").count;
+    const current_wealth = resourceDataState.find(resource => resource.resource_name == "wealth").count;
+
+    function handleErrors(errors) {
+        console.warn(errors);
+    }
+    const getBuildingUpkeep = (building) => {
+        let materials_cost = 0;
+        let food_cost = 0;
+        let wealth_cost = 0;
+        building.upgrade_sets.forEach((upgrade_set) => {
+            upgrade_set.upgrades.forEach((upgrade) => {
+                if (upgrade.status == "active" || upgrade.status == "activating") {
+                    materials_cost += upgrade.materials_upkeep;
+                    food_cost += upgrade.food_upkeep;
+                    wealth_cost += upgrade.wealth_upkeep;
+                }
+            });
+        });
+        return {
+            materials: materials_cost,
+            food: food_cost,
+            wealth: wealth_cost,
+        };
+    }
+    function romanize(num) {
+        switch (num) {
+            case 1:
+                return "I";
+            case 2:
+                return "II";
+            case 3:
+                return "III";
+            case 4:
+                return "IV";
+            case 5:
+                return "V";
+            case 6:
+                return "VI";
+            case 7:
+                return "VII";
+            case 8:
+                return "VIII";
+            case 9:
+                return "IX";
+            case 10:
+                return "X";
+            default:
+                return "I";
+        }
+    }
```

Comment:
This doesn't rely on any of the component's state or props, let's move it outside the component